Extension/src/api.ts

```diff
@@ -0,0 +1,110 @@
+     * to request configurations from the provider.
+     * @param provider An instance of the @see CustomConfigurationProvider instance representing the
+     * provider extension.
+     */
+    registerCustomConfigurationProvider(provider: CustomConfigurationProvider): void;
+    /**
+     * Notifies cpptools that the current configuration has changed. Upon receiving this notification,
+     * cpptools will request for the new configurations.
+     * @param provider An instance of the CustomConfigurationProvider instance representing the
+     * provider extension.
+     */
+    didCustomConfigurationChange(provider: CustomConfigurationProvider): void;
+}
+
+/**
+ * An interface to allow this extension to communicate with Custom Configuration Provider extensions
+ */
+export interface CustomConfigurationProvider {
+    /**
+     * The name of the Custom Configuration Provider extension
+     */
+    name: string;
+
+    /**
+     * A request to determine whether this provider can provide IntelliSense configurations for the given document.
+     * @param uri The URI of the document
+     * @param token Optional - The cancellation token
+     * @returns 'true' if this provider can provide IntelliSense configurations for the given document.
+     */
+    canProvideConfiguration(uri: vscode.Uri, token?: CancellationToken): Thenable<boolean>;
+
+    /**
+     * A request to get Intellisense configurations for the given files.
+     * @param uris A list of one of more URIs for the files to provide configurations for
+     * @param token Optional - The cancellation token
+     * @returns An list of @see SourceFileConfigurationItem for the documents that this provider is able to provide IntelliSense
+     * configurations for.
+     * Note: If this provider cannot provide configurations for a file in @param uris, then the file will not be included
+     * in the return value. An empty list will be returned if the provider cannot provide configurations for any of the files.
+     */
+    provideConfigurations(uris: vscode.Uri[], token?: CancellationToken): Thenable<SourceFileConfigurationItem[]>;
+}
+
+/**
+ * The model representing the custom IntelliSense configurations for a source file.
+ */
+export interface SourceFileConfiguration {
+    /**
+     * This must also include the system include path (compiler defaults) unless
+     * @param compilerPath is specified
+     */
+    includePath: string[];
+
+    /**
+     * This must also include the compiler default defines (__cplusplus, etc) unless
+     * @param compilerPath is specified
+     */
+    defines: string[];
+
+    /**
+     * "msvc-x64" or "clang-x64"
+     */
+    intelliSenseMode: string;
+    
+    /**
+     * The C or C++ standard
+     */
+    standard: string;
+
+    /**
+     * Any files that need to be included before the source file is parsed
+     */
+    forcedInclude?: string[];
+
+    /**

```

Comment:
@ThadHouse @bobbrow does this address the request?

Reply:
Does the low level language service cache values for the same compiler path? 

This looks great, but I would worry about performance if it checked the compiler for every file, even with the same path.

Reply:
👍

Reply:
@ThadHouse, yes the language service will cache values to avoid invoking the compiler repeatedly. I am working on this change right now.

Reply:
One more question. The normal docs say skip the headers and the compilerPath is msvc mode is set. Is that still the case with this API?

Reply:
Which docs? Clang for Windows also requires msvc mode, so you should specify the compilerPath in all cases where you want the extension to add system includes and defines for you.  We don't specifically query cl.exe in msvc mode yet since the mechanism isn't readily available (we read the registry instead), but we will in the future.

Reply:
https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/FAQ.md

Specifically this one, where if you want to use MSVC it says don't provide a compilerPath, and only add a compilerPath if using clang for windows.